rust/src/io/deletion.rs

```diff
@@ -0,0 +1,432 @@
+const BITMAP_THRESDHOLD: usize = 5_000;
+// TODO: Benchmark to find a better value.
+
+/// Represents a set of deleted row ids in a single fragment.
+#[derive(Debug, Clone)]
+pub(crate) enum DeletionVector {
+    NoDeletions,
+    Set(HashSet<u32>),
+    Bitmap(RoaringBitmap),
+}
+
+impl PartialEq for DeletionVector {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (DeletionVector::NoDeletions, DeletionVector::NoDeletions) => true,
+            (DeletionVector::Set(set1), DeletionVector::Set(set2)) => set1 == set2,
+            (DeletionVector::Bitmap(bitmap1), DeletionVector::Bitmap(bitmap2)) => {
+                bitmap1 == bitmap2
+            }
+            (DeletionVector::Set(set), DeletionVector::Bitmap(bitmap))
+            | (DeletionVector::Bitmap(bitmap), DeletionVector::Set(set)) => {
+                let set = set.iter().copied().collect::<RoaringBitmap>();
+                set == *bitmap
+            }
+            _ => false,
+        }
+    }
+}
+
+// TODO: impl methods for DeletionVector
+/// impl DeletionVector {
+///     pub fn get(i: u32) -> bool { ... }
+/// }
+/// impl BitAnd for DeletionVector { ... }
+
+impl IntoIterator for DeletionVector {
+    type IntoIter = Box<dyn Iterator<Item = Self::Item>>;
+    type Item = u32;
+
+    fn into_iter(self) -> Self::IntoIter {
+        match self {
+            DeletionVector::NoDeletions => Box::new(std::iter::empty()),
+            DeletionVector::Set(set) => Box::new(set.into_iter()),
+            DeletionVector::Bitmap(bitmap) => Box::new(bitmap.into_iter()),
+        }
+    }
+}
+
+impl FromIterator<u32> for DeletionVector {
+    fn from_iter<T: IntoIterator<Item = u32>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        match iter.size_hint() {
+            (_, Some(0)) => DeletionVector::NoDeletions,
+            (lower, _) if lower >= BITMAP_THRESDHOLD => {
+                let bitmap = iter.collect::<RoaringBitmap>();
+                DeletionVector::Bitmap(bitmap)
+            }
+            (_, Some(upper)) if upper < BITMAP_THRESDHOLD => {
+                let set = iter.collect::<HashSet<_>>();
+                DeletionVector::Set(set)
+            }
+            _ => {
+                // We don't know the size, so just try as a set and move to bitmap
+                // if it ends up being big.
+                let set = iter.collect::<HashSet<_>>();
+                if set.len() > BITMAP_THRESDHOLD {
+                    let bitmap = set.into_iter().collect::<RoaringBitmap>();
+                    DeletionVector::Bitmap(bitmap)
+                } else {
+                    DeletionVector::Set(set)
+                }
+            }
+        }
+    }
+}
+
+/// Get the Arrow schema for an Arrow deletion file.
+fn deletion_arrow_schema() -> Arc<Schema> {
+    Arc::new(Schema::new(vec![Field::new(
+        "row_id",
+        DataType::UInt32,
```

Comment:
potential tangent: maybe we want 64 bit uint to be safe? Just so that we don't have a "small" limit of number of rows we can store.

Reply:
So this isn't well documented, but the row ids elsewhere are a concatenation of the fragment id and the within-fragment row id (which is the position / offset of the row in the file).

https://github.com/lancedb/lance/blob/505dbcd1f186a9cfef364c71855c3f23d0dcee15/rust/src/io/reader.rs#L120-L123

So I think we already have a limit on the number of rows we can store per file based on this.

Reply:
makes sense. thank you for explaining!